companion/lib/Service/Satellite.js

```diff
@@ -292,31 +300,41 @@ class ServiceSatellite extends ServiceBase {
 			socket.write(`KEY-PRESS ERROR MESSAGE="Missing DEVICEID"\n`)
 			return
 		}
-		if (!params.KEY) {
-			socket.write(`KEY-PRESS ERROR DEVICEID="${params.DEVICEID}" MESSAGE="Missing KEY"\n`)
+		const id = `${params.DEVICEID}`
+		const device = this.#devices.get(id)
+		if (device === undefined || device.socket !== socket) {
+			socket.write(`KEY-PRESS ERROR DEVICEID="${id}" MESSAGE="Device not found"\n`)
 			return
 		}
-		if (!params.PRESSED) {
-			socket.write(`KEY-PRESS ERROR DEVICEID="${params.DEVICEID}" MESSAGE="Missing PRESSED"\n`)
+		if (!params.KEY) {
+			socket.write(`KEY-PRESS ERROR DEVICEID="${id}" MESSAGE="Missing KEY"\n`)
 			return
 		}
-
-		const key = Number(params.KEY)
-		if (isNaN(key) || key > LEGACY_MAX_BUTTONS || key < 0) {
-			socket.write(`KEY-PRESS ERROR DEVICEID="${params.DEVICEID}" MESSAGE="Invalid KEY"\n`)
+		if (!params.PRESSED) {
+			socket.write(`KEY-PRESS ERROR DEVICEID="${id}" MESSAGE="Missing PRESSED"\n`)
 			return
 		}
-
 		const pressed = !isFalsey(params.PRESSED)
 
-		const id = `${params.DEVICEID}`
-		const device = this.#devices.get(id)
+		const keynum = Number(params.KEY)
+		const keyParse = params.KEY.toString().match(/^\+?(\d+)\/\+?(\d+)$/)
```

Comment:
It would be nice to make a helper method which does this parsing of KEY. Perhaps it should always return `[x: number, y: number] | null` as well as doing the validation.
Then `doRotate` and `doButton` can be changed to accept that xy value instead of parsing it themselves.